From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul Gofman <pgofman@codeweavers.com>
Date: Tue, 27 Sep 2022 11:30:06 -0500
Subject: [PATCH] kernelbase: HACK: Do not expose version info for builtin DLLs
 for VC redists.

So VC Runtime 2013 redist actually installs the libraries.

CW-Bug-Id: #21343
Valve-commit: 7a0410d4aec28d849512ad6ad6c62fbd03b08e53
Picked: clean
Class: valve
Evidence: docs/evidence/2026-09-28-wine-proton-rebase.md
Offered-upstream: no
---
 dlls/kernelbase/version.c | 26 ++++++++++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/dlls/kernelbase/version.c b/dlls/kernelbase/version.c
index 5e59cdd..631fd27 100644
--- a/dlls/kernelbase/version.c
+++ b/dlls/kernelbase/version.c
@@ -785,6 +785,32 @@ DWORD WINAPI GetFileVersionInfoSizeExW( DWORD flags, LPCWSTR filename, LPDWORD r
     if ((hModule = LoadLibraryExW( filename, 0, LOAD_LIBRARY_AS_IMAGE_RESOURCE )))
     {
         HRSRC hRsrc = NULL;
+
+        static const char builtin_signature[] = "Wine builtin DLL";
+        HMODULE mod = (HMODULE)((ULONG_PTR)hModule & ~(ULONG_PTR)3);
+        char *signature = (char *)((IMAGE_DOS_HEADER *)mod + 1);
+        WCHAR exe_name[MAX_PATH];
+        IMAGE_NT_HEADERS *nt;
+        DWORD exe_name_len;
+
+        if ((exe_name_len = GetModuleFileNameW( NULL, exe_name, ARRAY_SIZE(exe_name) ))
+            && ((exe_name_len >= 16
+                && (!memcmp( exe_name + exe_name_len - 16, L"vcredist_x64.exe", 16 * sizeof(*exe_name) )
+                || !memcmp( exe_name + exe_name_len - 16, L"vcredist_x86.exe", 16 * sizeof(*exe_name) )))
+                || (exe_name_len >= 17
+                && (!memcmp( exe_name + exe_name_len - 17, L"vc_redist.x64.exe", 17 * sizeof(*exe_name) )
+                || !memcmp( exe_name + exe_name_len - 17, L"vc_redist.x86.exe", 17 * sizeof(*exe_name) )
+                || !memcmp( exe_name + exe_name_len - 17, L"VC_redist.x64.exe", 17 * sizeof(*exe_name) )
+                || !memcmp( exe_name + exe_name_len - 17, L"VC_redist.x86.exe", 17 * sizeof(*exe_name) ))))
+            && (nt = RtlImageNtHeader( mod )) && (char *)nt - signature >= sizeof(builtin_signature)
+            && !memcmp( signature, builtin_signature, sizeof(builtin_signature) ))
+        {
+            ERR("HACK: not exposing version info.\n");
+            FreeLibrary( hModule );
+            SetLastError( ERROR_RESOURCE_NAME_NOT_FOUND );
+            return 0;
+        }
+
         if (!(flags & FILE_VER_GET_LOCALISED))
         {
             LANGID english = MAKELANGID( LANG_ENGLISH, SUBLANG_DEFAULT );
